### PR TITLE
Feature/issue 3422 gsi timeout

### PR DESCRIPTION
**Pull request by tleyden**
Improves situation for #3422 

## Changes

1. Add `SG_TEST_DROP_INDEXES` env variable, which will cause all GSI indexes to be dropped prior to flushing the bucket if it's set to `true`
1. Update Jenkins job to take `SG_TEST_DROP_INDEXES` as a param
1. Make the test timeout configurable by an env variable `SG_TEST_TIMEOUT`
1. Update Jenkins job to take `SG_TEST_TIMEOUT` as a param
1. Modify `waitForSequence()` to take the wait time as a param.  This was useful to make it easier to see N1QL query timeouts which take up to 10s, and would otherwise be masked by the 2s `waitForSequence()` timeout higher up the stack
1. Submit parallel requests to Couchbase Server to drop indexes, in attempt to speed it up (not sure if it did)
1. `TestUserAccessRace()` only runs against Walrus
1. A few tests don't need to flush the bucket (or drop indexes), and so they were updated to have `RestTester.NoFlush=true` in attempt to speed things up

## Test performance running full suite

On an AWS `m4.4xlarge` instance, it's taking approximately 1.5 - 3 hours to run the full suite w/  `SG_TEST_DROP_INDEXES` and all tests passed.   Test output: https://gist.github.com/b7a75f63884d644791d2c0a70992f4c2

## Integration tests

http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/604/  (SG_TEST_DROP_INDEXES=true)
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/605/ (SG_TEST_DROP_INDEXES=false)

### CODE REVIEW

**Comment by tleyden** [comment]
> http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/604/ (SG_TEST_DROP_INDEXES=true)

This took 6 hours, and failed with https://github.com/couchbase/sync_gateway/issues/3539#issuecomment-387810263

> http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/605/ (SG_TEST_DROP_INDEXES=false)

Had failures, which I'm assuming are expected due to not dropping the GSI indexes.  I wish this was easier to tell though.  @adamcfraser what do you think of bumping `DefaultWaitForSequenceTesting` (new w/ this PR) to 30s, in the hopes of seeing n1ql query timeouts in the logs?  Actually it might require another logging change, since I don't remember if the n1ql query timeouts make it into the logs by default .. but maybe individual tests could tune the logging accordingly.

**Comment by tleyden** [comment]
> http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/604/ (SG_TEST_DROP_INDEXES=true)

It passed on a re-run here:

http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/610/parameters/

File: base/bucket_n1ql.go [hunk 1]
```diff
@@ -48,7 +48,11 @@ func (bucket *CouchbaseBucketGoCB) Query(statement string, params interface{}, c
 
 	waitTime := 10 * time.Millisecond
 	for i := 1; i <= MaxQueryRetries; i++ {
+
+		Tracef(KeyQuery, "Executing N1QL query: %v", UD(n1qlQuery))
 		queryResults, queryErr := bucket.ExecuteN1qlQuery(n1qlQuery, params)
+		Tracef(KeyQuery, "Done executing N1QL query: %v.  Error: %v", UD(n1qlQuery), queryErr)
```
Comment (adamcfraser) [inline base/bucket_n1ql.go]: Can we remove this logging (or at least only log in the queryErr != nil case?)

File: base/util_testing.go [hunk 2]
```diff
@@ -173,13 +172,28 @@ func TestUseXattrs() bool {
 	useXattrs := os.Getenv(TestEnvSyncGatewayUseXattrs)
 	switch {
 	case strings.ToLower(useXattrs) == strings.ToLower(TestEnvSyncGatewayTrue):
-		Infof(KeyAll, "Using xattrs: strings.ToLower(useXattrs) == strings.ToLower(TestEnvSyncGatewayTrue).  |%v| == |%v|", strings.ToLower(useXattrs), strings.ToLower(TestEnvSyncGatewayTrue))
 		return true
 	default:
-		Infof(KeyAll, "NOT Using xattrs: strings.ToLower(useXattrs) != strings.ToLower(TestEnvSyncGatewayTrue).  |%v| != |%v|", strings.ToLower(useXattrs), strings.ToLower(TestEnvSyncGatewayTrue))
+		// Otherwise fallback to hardcoded default
+		return DefaultUseXattrs
```
Comment (adamcfraser) [inline base/util_testing.go]: Good idea to clean this up, but it looks like we can go a bit further and convert to a simple if statement.

File: base/util_testing.go [hunk 3]
```diff
@@ -173,13 +172,28 @@ func TestUseXattrs() bool {
...
+
+	// First check if the SG_TEST_USE_XATTRS env variable is set
+	dropIndexes := os.Getenv(TestEnvSyncGatewayDropIndexes)
+	switch {
+	case strings.ToLower(dropIndexes) == strings.ToLower(TestEnvSyncGatewayTrue):
+		return true
+	default:
+		// Otherwise fallback to hardcoded default
+		return DefaultDropIndexes
+	}
```
Comment (adamcfraser) [inline base/util_testing.go]: Same as above - is there a reason this should be a switch?

File: base/util_testing.go [hunk 4]
```diff
@@ -235,56 +248,115 @@ func (tbm *TestBucketManager) OpenTestBucket() (bucketExists bool, err error) {
...
+	select {
+	case asyncError := <-asyncErrors:
+		return asyncError
+	default:
+	}
+
+	return nil
+}
+
+// Get a list of all index names in the bucket
```
Comment (adamcfraser) [inline base/util_testing.go:319]: Might be a good time to replace this w/ gocb's BucketManager.GetIndexes().

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: base/bucket_n1ql.go (hunk 1 above)
Comment (tleyden) [reply base/bucket_n1ql.go]: Removed

File: base/util_testing.go (hunk 2 above)
Comment (tleyden) [reply base/util_testing.go]: Fixed

File: base/util_testing.go (hunk 3 above)
Comment (tleyden) [reply base/util_testing.go]: Fixed

File: base/util_testing.go (hunk 4 above)
Comment (tleyden) [reply base/util_testing.go:319]: Thanks for catching. Fixed

**Comment by tleyden** [comment]
Pushed up changes, ready for another pass

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**